includes/admin/extensions/abstract-extension.php

```diff
@@ -0,0 +1,209 @@
+		if ( $this->is_activated() ) {
+			return;
+		}
+		$this->do_single_extension_card();
+	}
+
+	/**
+	 * Outputs a single extension card.
+	 *
+	 * @since 2.11.x
+	 * @param bool|int $item_id Optional: the individual extension product ID.
+	 * @return void
+	 */
+	public function do_single_extension_card( $item_id = false ) {
+		$config       = $this->get_configuration( $item_id );
+		$product_data = $this->get_product_data( $item_id );
+		if ( ! $product_data ) {
+			return;
+		}
+		$this->manager->do_extension_card(
+			$product_data,
+			$config,
+			$this->get_button_parameters( $config, $product_data ),
+			$this->get_link_parameters( $config, $product_data->info->title )
+		);
+	}
+
+	/**
+	 * Gets the product data for a specific extension.
+	 *
+	 * @param boolean $item_id
+	 * @return bool|object False if there is no data; API object if there is.
+	 */
+	public function get_product_data( $item_id = false ) {
+		$body = $this->get_api_body();
+		if ( empty( $body ) ) {
+			return false;
+		}
+		require_once EDD_PLUGIN_DIR . 'includes/admin/extensions/class-extensions-api.php';
+		$api          = new ExtensionsAPI();
+		$product_data = $api->get_product_data( $body, $this->item_id );
+		if ( ! $product_data ) {
+			return false;
+		}
+
+		return $this->item_id ? $product_data : $product_data[ $item_id ];
+	}
+
+	/**
+	 * Gets the array for the body of the API request.
+	 * Classes may need to override this (for example, to query a specific tag).
+	 *
+	 * @return array
+	 */
+	protected function get_api_body() {
+		return array(
+			'product' => $this->item_id,
+		);
+	}
+
+	/**
+	 * Gets the type for the button data-type attribute.
+	 * This is intended to sync with the Products API request.
+	 * Default is product.
+	 *
+	 * Really a shim for array_key_first.
+	 *
+	 * @param array $array
+	 * @return void
+	 */
+	private function get_type( array $array ) {
+		if ( function_exists( 'array_key_first' ) ) {
+			return array_key_first( $array );
+		}
+		foreach ( $array as $key => $unused ) {
+			return $key;
+		}
+
+		return 'product';
+	}
+
+	/**
+	 * Gets the button parameters.
+	 * Classes should not need to replace this method.
+	 *
+	 * @param array  $config       The array of provided data about the extension.
+	 * @param object $product_data The extension data returned from the Products API.
+	 * @return array
+	 */
+	protected function get_button_parameters( $config, $product_data ) {
+		$item_id = ! empty( $product_data->info->id ) ? $product_data->info->id : $this->item_id;
+		$body    = $this->get_api_body();
+		$type    = $this->get_type( $body );
+		$id      = $body[ $type ];
+		$button  = array(
+			'type'    => $type,
+			'id'      => $id,
+			'product' => $item_id,
+		);
+		// If the extension is not installed, the button will prompt to install and activate it.
+		if ( ! $this->manager->is_plugin_installed( $config['basename'] ) ) {
+			$download_url = $this->manager->get_download_url( $item_id, 'extension' );
```

Comment:
We need to rework this. The new "Email Marketing" tab is an example. It's doing a separate API request for every single extension that's not already installed. This is going to 1) impact our site's performance; and 2) impact the performance of actually loading that page.

I think a better method is to actually fetch the download URL as part of the ajax callback (so only after the button is clicked). So when you click an "Install & Activate" button, the first part of that ajax callback can be to get the download URL, then show an error if it cannot be retrieved for some reason.

Reply:
Oh, that's a great idea--and yes I was concerned about that. I can work on that.